cep-recipe-jinja.md

```diff
@@ -0,0 +1,320 @@
+
+You can also check for the existence of an environment variable:
+
+- `${{ env.exists("MY_ENV_VAR") }}` will return a boolean `true` if the environment variable `MY_ENV_VAR` is set and `false` otherwise.
+
+## Jinja filters
+
+A feature of `jinja` is called "filters". Filters are functions that can be applied to variables in a template expression.
+
+The syntax for a filter is `{{ variable | filter_name }}`. A filter can also take arguments, such as `... | replace('foo', 'bar')`.
+
+The following Jinja filters are available:
+
+- `replace`: replace a string with another string (e.g. `"{{ 'foo' | replace('oo', 'aa') }}"` will return `"faa"`)
+- `lower`: convert a string to lowercase (e.g. `"{{ 'FOO' | lower }}"` will return `"foo"`)
+- `upper`: convert a string to uppercase (e.g. `"{{ 'foo' | upper }}"` will return `"FOO"`)
+- `int`: convert a string to an integer (e.g. `"{{ '42' | int }}"` will return `42`)
+
+There are more filters available as [documented by MiniJinja](https://docs.rs/minijinja/latest/minijinja/filters/index.html).
+
+> [!NOTE]
+> Should we add all filters from `minijinja` to this spec? Probably ...
+
+### Extra filters for recipes
+
+#### The `version_to_buildstring` filter
+
+- `${{ python | version_to_buildstring }}` converts a version from the variant to a build string (it removes the `.` character and takes only the first two elements of the version).
+For example the following 
+
+```yaml
+context:
+  cuda: "11.2.0"
+
+build:
+  string: ${{ hash }}_cuda${{ cuda_version | version_to_buildstring }}
+```
+
+Would evaluate to a `abc123_cuda112` (assuming the hash was `abc123`).
+
+### Various remarks
+
+#### Inline conditionals with Jinja
+
+The new recipe format allows for inline conditionals with Jinja. If they are falsey, and no `else` branch exists, they will render to an empty string (which is, for example in a list or dictionary, equivalent to a YAML `null`).
+
+When we render the recipe, we filter all values that are `null` from the YAML output. This allows us to write and parse the following easily:
+
+```yaml
+requirements:
+  host:
+    - ${{ "numpy" if cuda == "yes" }}
+```
+
+If `cuda` is not equal to yes, the first item of the host requirements will be empty (null) and thus filtered from the final list.
+
+This also works for dictionary keys:
+
+```yaml
+build:
+  number: ${{ 100 if cuda == "yes" }}
+  # or an `else` branch can be used, of course
+  number: ${{ 100 if cuda == "yes" else 0 }}
+```
+
+---
+
+### Things to decide
+
+- Limit the filters from Jinja to a default subset?
+- Should `env` work more like Python `env`
+  - `env["FOO"]` instead of `env.get("FOO")`
+  - `env.get("FOO", "default")` instead of `env.get_default("FOO", "default")`
+- Or should it be more like Github Actions
+  - `env.FOO` instead of `env.get("FOO")`
+  - `env.FOO or "default"` for default values
+- Should `${{ compiler('c') }}` evaluate directly to `gcc_linux-64` or should there be an intermediate representation (as is now)
```

Comment:
Isn't this an implementation detail of the tool? In the sense that even if you have an intermediate representation, the end result will be `gcc_linux-64 11.1.0` right?

Reply:
Yep, it's just something on the tool side. It's gone now.